dump/src/reader/v1/mod.rs

```diff
@@ -1,173 +1,265 @@
+        V1IndexReader::new(&self.path, self.metadata.clone())
+    }
+
+    pub fn metadata(&self) -> &IndexMetadata {
+        &self.metadata.metadata
+    }
+}
+
+pub struct V1IndexReader {
+    metadata: IndexMetadataV1,
     documents: BufReader<File>,
     settings: BufReader<File>,
     updates: BufReader<File>,
-
-    current_update: Option<UpdateStatus>,
 }
 
 impl V1IndexReader {
-    pub fn new(name: String, path: &Path) -> Result<Self> {
-        let mut ret = V1IndexReader {
-            name,
+    pub fn new(path: &Path, metadata: IndexMetadataV1) -> Result<Self> {
+        Ok(V1IndexReader {
+            metadata,
             documents: BufReader::new(File::open(path.join("documents.jsonl"))?),
             settings: BufReader::new(File::open(path.join("settings.json"))?),
             updates: BufReader::new(File::open(path.join("updates.jsonl"))?),
-            current_update: None,
-        };
-        ret.next_update();
+        })
+    }
+
+    pub fn metadata(&self) -> &IndexMetadata {
+        &self.metadata.metadata
+    }
 
-        Ok(ret)
+    pub fn documents(&mut self) -> Result<impl Iterator<Item = Result<Document>> + '_> {
+        Ok((&mut self.documents)
+            .lines()
+            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) }))
     }
 
-    pub fn next_update(&mut self) -> Result<Option<UpdateStatus>> {
-        let current_update = if let Some(line) = self.updates.lines().next() {
-            Some(serde_json::from_str(&line?)?)
-        } else {
-            None
-        };
+    pub fn settings(&mut self) -> Result<self::settings::Settings> {
+        Ok(serde_json::from_reader(&mut self.settings)?)
+    }
 
-        Ok(std::mem::replace(&mut self.current_update, current_update))
+    pub fn tasks(self) -> impl Iterator<Item = Result<Task>> {
+        self.updates.lines().map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) })
     }
 }
 
 impl V1Reader {
     pub fn open(dump: TempDir) -> Result<Self> {
-        let mut meta_file = fs::read(dump.path().join("metadata.json"))?;
-        let metadata = serde_json::from_reader(&*meta_file)?;
+        let meta_file = fs::read(dump.path().join("metadata.json"))?;
+        let metadata: Metadata = serde_json::from_reader(&*meta_file)?;
 
         let mut indexes = Vec::new();
 
-        let entries = fs::read_dir(dump.path())?;
-        for entry in entries {
-            let entry = entry?;
-            if entry.file_type()?.is_dir() {
-                indexes.push(V1IndexReader::new(
-                    entry
-                        .file_name()
-                        .to_str()
-                        .ok_or(Error::BadIndexName)?
-                        .to_string(),
-                    &entry.path(),
-                )?);
-            }
+        for index in metadata.indexes.into_iter() {
+            let index_path = dump.path().join(&index.uid);
+            indexes.push(V1Index::new(index_path, index));
```

Comment:
From what I see, the indexes are actually ordered in the dump, right?

Reply:
yes, that is the case: in the `metadata.json` of the dumps v1, we have a list of indexes, so it is definitely ordered.

I guess I copy pasted this from v2 where indexes weren't ordered